### PR TITLE
[Review] Prefer user supplied parameter for remove and deploy

### PR DESCRIPTION
**Pull request by ericstoekl**
Prefer user supplied parameter for remove and deploy

Signed-off-by: Eric Stoekl <ems5311@gmail.com>

## Description

This change is to prefer the user-supplied function name when performing the `faas-cli deploy` and `faas-cli remove` operations. Currently, if a `stack.yml` file is present in your working directory (which is designated by `faas-cli` as the default filename for a function stack), and you specify a function name (e.g. `faas-cli remove myFunc`), the contents of the `stack.yml` file will be operated on. This is unpredictable behavior and may be confusing to some users.

## Motivation and Context

- [X] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
https://github.com/openfaas/faas-cli/issues/280

## How Has This Been Tested?

Tests were created for `deploy` and `remove` to ensure proper functionality. Tests were created prior to making the changes to the business logic.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: commands/remove.go [hunk 1]
```diff
@@ -40,7 +40,12 @@ func runDelete(cmd *cobra.Command, args []string) error {
 	var services stack.Services
 	var gatewayAddress string
 	var yamlGateway string
-	if len(yamlFile) > 0 {
+
+	argsSelection := false
```
Comment (itscaro) [inline commands/remove.go]: maybe rename to `hasArgs`?

File: commands/remove.go [hunk 2]
```diff
@@ -40,7 +40,12 @@ func runDelete(cmd *cobra.Command, args []string) error {
...
+
+	if len(args) == 1 {
```
Comment (alexellis) [inline commands/remove.go]: What if you have both args and YAML?

Comment (ericstoekl) [reply commands/remove.go]: It prefers args and does not perform the operation on the `stack.yml`. I think that is more expected by the user.

**Comment by ericstoekl** [comment]
Squash merge to rename args to `hasArgs` (remove.go) and `hasFuncName` (deploy.go)

File: commands/deploy.go [hunk 3]
```diff
@@ -128,6 +132,8 @@ func RunDeploy(
 		if parsedServices != nil {
 			services = *parsedServices
 		}
+	} else {
+		return fmt.Errorf("please provide a --name for your function as it will be deployed on FaaS")
```
Comment (alexellis) [inline commands/deploy.go]: `FaaS` should be `OpenFaaS`

Can we re-write the message? It doesn't sound very natural.

File: commands/deploy_test.go [hunk 4]
```diff
@@ -90,6 +91,60 @@ func Test_deploy(t *testing.T) {
 		faasCmd.Execute()
 	})
 
+	if found, err := regexp.MatchString(`(?m:Deploying: test-function.)`, stdOut); err != nil || !found {
+		t.Fatalf("Output is not as expected:\n%s", stdOut)
```
Comment (alexellis) [inline commands/deploy_test.go]: These messages are hard to work with when they appear in an error log.

Can you please include the desired regex in the message? i.e. "Tried to match regex '%s' but couldn't, got: %s"

File: commands/remove.go [hunk 5]
```diff
@@ -40,7 +40,12 @@ func runDelete(cmd *cobra.Command, args []string) error {
 	var services stack.Services
 	var gatewayAddress string
 	var yamlGateway string
-	if len(yamlFile) > 0 {
+
+	hasArgs := false
+
+	if len(args) == 1 {
```
Comment (alexellis) [inline commands/remove.go]: Finding the code that uses `hasArgs = true` hard to read.. can we make this simpler?

File: commands/remove_test.go [hunk 6]
```diff
@@ -22,10 +24,60 @@ func Test_remove(t *testing.T) {
...
+	stdOut := test.CaptureStdout(func() {
+		faasCmd.SetArgs([]string{
+			"remove",
+			"--gateway=" + s.URL,
+			"test-function",
+		})
+		faasCmd.Execute()
+	})
+
+	expectedStdOut := "Deleting: test-function."
```
Comment (alexellis) [inline commands/remove_test.go]: why do we need a variable `expectedStdOut`?

**Comment by alexellis** [comment]
@ericstoekl there's several review comments which are now over a month old. Do you want me to close your PR or are you able to pick it up again?

File: commands/remove_test.go (hunk 6 above)
Comment (ericstoekl) [reply commands/remove_test.go]: It helps for printing out the error message.

**Comment by ericstoekl** [comment]
@alexellis Addressed the comments and synced+rebased

**Comment by rgee0** [comment]
Late to the party, but could this follow a similar pattern as the gateway url work in https://github.com/openfaas/faas-cli/blob/master/commands/config_priorty.go

**Comment by ericstoekl** [comment]
Rebased but I like Richard's idea quite a bit -- I will update and make it similar to the `config_priority.go` example.

**Comment by ericstoekl** [comment]
Updated the PR to include the `getFunctionSource` function which returns the source of the function definition.

**Comment by alexellis** [comment]
I'll look at the latest change now

File: commands/build.go [hunk 7]
```diff
@@ -127,16 +137,9 @@ func runBuild(cmd *cobra.Command, args []string) error {
...
 
-	if pullErr := PullTemplates(DefaultTemplateRepository); pullErr != nil {
-		return fmt.Errorf("could not pull templates for OpenFaaS: %v", pullErr)
-	}
-
-	if len(services.Functions) > 0 {
 		build(&services, parallel, shrinkwrap)
-
-	} else {
+	} else if functionSource == "inline" {
```
Comment (alexellis) [inline commands/build.go]: If we're going to use hard-coded strings `"inline"` and `"YAML"` then these should be abstracted away in constants. I would call it:

```
const argumentSource = "arguments"
const  YAMLSource = "yamlFile"
```

For instance.

File: commands/build.go [hunk 8]
```diff
@@ -118,7 +118,17 @@ func parseBuildArgs(args []string) (map[string]string, error) {
...
 
 	var services stack.Services
-	if len(yamlFile) > 0 {
+
+	functionSource, err := getFunctionSource(functionName)
+	if err != nil {
+		return err
+	}
+
+	if pullErr := PullTemplates(DefaultTemplateRepository); pullErr != nil {
```
Comment (alexellis) [inline commands/build.go]: Is this new? Was it added by this PR?

Comment (rgee0) [reply commands/build.go]: not new - line 132 in the original

File: commands/build.go (hunk 7 above)
Comment (rgee0) [reply commands/build.go]: Agree with Alex's comment.  The strings are also used in too many places not to be constants.

**Comment by ericstoekl** [comment]
Changes posted -- PTAL @alexellis @rgee0

**Comment by rgee0** [comment]
Changes addressed.  Lets move to test.  
@ericstoekl can we get the community to checkout this PR and test it locally and report back on what was tested and the results?

**Comment by rgee0** [comment]
Derek add label: status/testing

**Comment by rgee0** [comment]
Hello @ericstoekl 

Just running through some testing on this and whilst I can remove using:

```
./faas-cli-darwin remove funcName
```
I cant remove using 

```
$ ./faas-cli-darwin remove  -f ./hello-python.yml
panic: runtime error: index out of range
goroutine 1 [running]:
github.com/openfaas/faas-cli/commands.runDelete(0x1606900, 0xc42007b7d0, 0x0, 0x3, 0x0, 0x0)
        /go/src/github.com/openfaas/faas-cli/commands/remove.go:45 +0x5f2
``

if I use 0.6.5 then it runs ok:

```
$ faas-cli remove  -f ./hello-python.yml
Deleting: hello-python.
Removing old function.
```

**Comment by ericstoekl** [comment]
@rgee0 Thanks for testing, I get the same issue. I'll fix this and update the PR.

**Comment by ericstoekl** [comment]
Updated the PR with a fix, @rgee0  PTAL

**Comment by ericstoekl** [comment]
Synced up the PR and rebased to master. In addition, I've created a test script you can run to make sure this code works.

Test this PR out by deploying openfaas on docker swarm on `localhost`, then `cd` into the `faas-cli/` repo directory and run the following:

```
curl https://gist.githubusercontent.com/ericstoekl/9bb4fa2f9f66a9ba068f519615a5ef6c/raw/22a48733d5591d1ac155131040e8d16eb0f9f9f4/argTest.sh > argTest.sh
sh argTest.sh
```

This will exercise the functionality of this PR, which is to honor the user-supplied parameters instead of just deploying/removing the functions in the `stack.yml` file, if one is present in your working directory. Please let me know the results of running the test script.

**Comment by alexellis** [comment]
Can you run it and give the results @ericstoekl ?

**Comment by ericstoekl** [comment]
@alexellis Sure:

```
Stack file updated: stack.yml
----------------
Running faas-cli 0.6.13 with deploy --image alpine --fprocess cat --name catsvc
----------------
Deploying: p1.

Deployed. 200 OK.
URL: http://127.0.0.1:8080/function/p1

Deploying: p2.

Deployed. 200 OK.
URL: http://127.0.0.1:8080/function/p2

----------------
Running faas-cli for PR 304 with deploy --image alpine --fprocess cat --name catsvc
----------------
Deploying: catsvc.
Function catsvc already exists, attempting rolling-update.

Deployed. 200 OK.
URL: http://127.0.0.1:8080/function/catsvc

----------------
Running faas-cli 0.6.13 with remove p1
----------------
Deleting: p1.
Removing old function.
Deleting: p2.
Removing old function.
----------------
Running faas-cli for PR 304 with remove p1
----------------
Deleting: p1.
No existing function to remove
```

**Comment by alexellis** [comment]
I'm worried about introducing more edge-cases. Is there another solution that may be simpler?

Could we refuse to run a command if a parameter is given and we're using stack.yml? Would that work?

Solution, user moves out the directory.

**Pull request closed without merging**